PgHasArrayType is not implemented for Vec<i64>

### Bug Description
after upgrading to `0.7.0` I'm getting this error when using https://docs.rs/sqlx/latest/sqlx/trait.Type.html#transparent
```
error[E0277]: the trait bound `Vec<i64>: PgHasArrayType` is not satisfied
 --> src/lib.rs:1:10
  |
1 | #[derive(sqlx::Type)]
  |          ^^^^^^^^^^ the trait `PgHasArrayType` is not implemented for `Vec<i64>`
  |
  = help: the trait `PgHasArrayType` is implemented for `Vec<u8>`
  = help: see issue #48214
  = note: this error originates in the derive macro `sqlx::Type` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `sqlx-test2` (lib) due to previous error
```

### Minimal Reproduction
```console
cargo init --lib sqlx-test
```
`src/lib.rs`:
```rust
#[derive(sqlx::Type)]
#[sqlx(transparent)]
pub struct ScriptHashes(pub Vec<i64>);
```
`Cargo.toml`:
```rust
[package]
name = "sqlx-test2"
version = "0.1.0"
edition = "2021"

[dependencies]
sqlx = { version = "0.7.0", features = [
    "macros",
    "postgres",
] }
```
then
```console
cargo build
```

### Info
* SQLx version: 0.7.0
* SQLx features enabled: `macros` and `postgres`
* Database server and version: Postgres 14 or 15
* Operating system: linux
* `rustc --version`: rustc 1.70.0 (90c541806 2023-05-31)

### additional info
* issue doesn't occur with version `0.6.3`
* I think that this might be related https://github.com/launchbadge/sqlx/pull/2086

--- a/sqlx-core/src/types/mod.rs
+++ b/sqlx-core/src/types/mod.rs
@@ -104,13 +104,42 @@ pub use json::{Json, JsonRawValue, JsonValue};
 ///
 /// ### Transparent
 ///
-/// Rust-only domain or wrappers around SQL types. The generated implementations directly delegate
+/// Rust-only domain wrappers around SQL types. The generated implementations directly delegate
 /// to the implementation of the inner type.
 ///
 /// ```rust,ignore
 /// #[derive(sqlx::Type)]
 /// #[sqlx(transparent)]
 /// struct UserId(i64);
+/// ```
+///
+/// ##### Note: `PgHasArrayType`
+/// If you have the `postgres` feature enabled, this derive also generates a `PgHasArrayType` impl
+/// so that you may use it with `Vec` and other types that decode from an array in Postgres:
+///
+/// ```rust,ignore
+/// let user_ids: Vec<UserId> = sqlx::query_scalar("select '{ 123, 456 }'::int8[]")
+///    .fetch(&mut pg_connection)
+///    .await?;
+/// ```
+///
+/// However, if you are wrapping a type that does not implement `PgHasArrayType`
+/// (e.g. `Vec` itself, because we don't currently support multidimensional arrays),
+/// you may receive an error:
+///
+/// ```rust,ignore
+/// #[derive(sqlx::Type)] // ERROR: `Vec<i64>` does not implement `PgHasArrayType`
+/// #[sqlx(transparent)]
+/// struct UserIds(Vec<i64>);
+/// ```
+///
+/// To remedy this, add `#[sqlx(no_pg_array)]`, which disables the generation
+/// of the `PgHasArrayType` impl:
+///
+/// ```rust,ignore
+/// #[derive(sqlx::Type)]
+/// #[sqlx(transparent, no_pg_array)]
+/// struct UserIds(Vec<i64>);
 /// ```
 ///
 /// ##### Attributes
@@ -121,6 +150,7 @@ pub use json::{Json, JsonRawValue, JsonValue};
 ///   given type is different than that of the inferred type (e.g. if you rename the above to
 ///   `VARCHAR`). Affects Postgres only.
 /// * `#[sqlx(rename_all = "<strategy>")]` on struct definition: See [`derive docs in FromRow`](crate::from_row::FromRow#rename_all)
+/// * `#[sqlx(no_pg_array)]`: do not emit a `PgHasArrayType` impl (see above).
 ///
 /// ### Enumeration
 ///

--- a/sqlx-macros-core/src/derives/attributes.rs
+++ b/sqlx-macros-core/src/derives/attributes.rs
@@ -56,6 +56,7 @@ pub struct SqlxContainerAttributes {
     pub type_name: Option<TypeName>,
     pub rename_all: Option<RenameAll>,
     pub repr: Option<Ident>,
+    pub no_pg_array: bool,
 }
 
 pub struct SqlxChildAttributes {
@@ -71,6 +72,7 @@ pub fn parse_container_attributes(input: &[Attribute]) -> syn::Result<SqlxContai
     let mut repr = None;
     let mut type_name = None;
     let mut rename_all = None;
+    let mut no_pg_array = None;
 
     for attr in input
         .iter()
@@ -86,6 +88,10 @@ pub fn parse_container_attributes(input: &[Attribute]) -> syn::Result<SqlxContai
                         NestedMeta::Meta(meta) => match meta {
                             Meta::Path(p) if p.is_ident("transparent") => {
                                 try_set!(transparent, true, value)
+                            }
+
+                            Meta::Path(p) if p.is_ident("no_pg_array") => {
+                                try_set!(no_pg_array, true, value);
                             }
 
                             Meta::NameValue(MetaNameValue {
@@ -148,6 +154,7 @@ pub fn parse_container_attributes(input: &[Attribute]) -> syn::Result<SqlxContai
         repr,
         type_name,
         rename_all,
+        no_pg_array: no_pg_array.unwrap_or(false),
     })
 }
 
@@ -229,6 +236,12 @@ pub fn check_enum_attributes(input: &DeriveInput) -> syn::Result<SqlxContainerAt
         input
     );
 
+    assert_attribute!(
+        !attributes.no_pg_array,
+        "unused #[sqlx(no_pg_array)]; derive does not emit `PgHasArrayType` impls for enums",
+        input
+    );
+
     Ok(attributes)
 }
 
@@ -285,6 +298,12 @@ pub fn check_struct_attributes<'a>(
     assert_attribute!(
         attributes.rename_all.is_none(),
         "unexpected #[sqlx(rename_all = ..)]",
+        input
+    );
+
+    assert_attribute!(
+        !attributes.no_pg_array,
+        "unused #[sqlx(no_pg_array)]; derive does not emit `PgHasArrayType` impls for custom structs",
         input
     );
 

--- a/sqlx-macros-core/src/derives/type.rs
+++ b/sqlx-macros-core/src/derives/type.rs
@@ -90,7 +90,7 @@ fn expand_derive_has_sql_type_transparent(
             }
         );
 
-        if cfg!(feature = "postgres") {
+        if cfg!(feature = "postgres") && !attr.no_pg_array {
             tokens.extend(quote!(
                 #[automatically_derived]
                 impl #array_impl_generics ::sqlx::postgres::PgHasArrayType for #ident #ty_generics

--- a/sqlx-postgres/src/types/array.rs
+++ b/sqlx-postgres/src/types/array.rs
@@ -9,6 +9,44 @@ use crate::types::Oid;
 use crate::types::Type;
 use crate::{PgArgumentBuffer, PgTypeInfo, PgValueFormat, PgValueRef, Postgres};
 
+/// Provides information necessary to encode and decode Postgres arrays as compatible Rust types.
+///
+/// Implementing this trait for some type `T` enables relevant `Type`,`Encode` and `Decode` impls
+/// for `Vec<T>`, `&[T]` (slices), `[T; N]` (arrays), etc.
+///
+/// ### Note: `#[derive(sqlx::Type)]`
+/// If you have the `postgres` feature enabled, `#[derive(sqlx::Type)]` will also generate
+/// an impl of this trait for your type if your wrapper is marked `#[sqlx(transparent)]`:
+///
+/// ```rust,ignore
+/// #[derive(sqlx::Type)]
+/// #[sqlx(transparent)]
+/// struct UserId(i64);
+///
+/// let user_ids: Vec<UserId> = sqlx::query_scalar("select '{ 123, 456 }'::int8[]")
+///    .fetch(&mut pg_connection)
+///    .await?;
+/// ```
+///
+/// However, this may cause an error if the type being wrapped does not implement `PgHasArrayType`,
+/// e.g. `Vec` itself, because we don't currently support multidimensional arrays:
+///
+/// ```rust,ignore
+/// #[derive(sqlx::Type)] // ERROR: `Vec<i64>` does not implement `PgHasArrayType`
+/// #[sqlx(transparent)]
+/// struct UserIds(Vec<i64>);
+/// ```
+///
+/// To remedy this, add `#[sqlx(no_pg_array)]`, which disables the generation
+/// of the `PgHasArrayType` impl:
+///
+/// ```rust,ignore
+/// #[derive(sqlx::Type)]
+/// #[sqlx(transparent, no_pg_array)]
+/// struct UserIds(Vec<i64>);
+/// ```
+///
+/// See [the documentation of `Type`][Type] for more details.
 pub trait PgHasArrayType {
     fn array_type_info() -> PgTypeInfo;
     fn array_compatible(ty: &PgTypeInfo) -> bool {

--- a/tests/postgres/derives.rs
+++ b/tests/postgres/derives.rs
@@ -10,6 +10,13 @@ use std::ops::Bound;
 #[sqlx(transparent)]
 struct Transparent(i32);
 
+#[derive(PartialEq, Debug, sqlx::Type)]
+// https://github.com/launchbadge/sqlx/issues/2611
+// Previously, the derive would generate a `PgHasArrayType` impl that errored on an
+// impossible-to-satisfy `where` bound. This attribute allows the user to opt-out.
+#[sqlx(transparent, no_pg_array)]
+struct TransparentArray(Vec<i64>);
+
 #[sqlx_macros::test]
 async fn test_transparent_slice_to_array() -> anyhow::Result<()> {
     let mut conn = new::<Postgres>().await?;
@@ -137,6 +144,11 @@ struct RangeInclusive(PgRange<i32>);
 test_type!(transparent<Transparent>(Postgres,
     "0" == Transparent(0),
     "23523" == Transparent(23523)
+));
+
+test_type!(transparent_array<TransparentArray>(Postgres,
+    "'{}'::int8[]" == TransparentArray(vec![]),
+    "'{ 23523, 123456, 789 }'::int8[]" == TransparentArray(vec![23523, 123456, 789])
 ));
 
 test_type!(weak_enum<Weak>(Postgres,

